test/upgrade/upgrade_suite_test.go

```diff
@@ -0,0 +1,152 @@
+	flag.StringVar(&supportBundleDir, "support-bundle-dir", "", "Support bundles are saved in this dir when specified.")
+	flag.StringVar(&cloudProvider, "cloud-provider", string(runtimev1alpha1.AzureCloudProvider),
+		"Cloud Provider. Default is Azure.")
+	flag.StringVar(&chartDir, "chart-dir", "../../build/charts/nephe", "Helm chart directory.")
+	rand.Seed(time.Now().Unix())
+}
+
+func TestUpgrade(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Upgrade Suite")
+}
+
+var _ = BeforeSuite(func() {
+	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), logging.UseDevMode()))
+
+	// required fields.
+	Expect(fromVersion).ToNot(BeEmpty())
+
+	var err error
+	By("Bootstrapping the test environment")
+	err = clientgoscheme.AddToScheme(scheme)
+	Expect(err).NotTo(HaveOccurred())
+
+	err = cloudv1alpha1.AddToScheme(scheme)
+	Expect(err).NotTo(HaveOccurred())
+
+	err = antreav1alpha1.AddToScheme(scheme)
+	Expect(err).NotTo(HaveOccurred())
+
+	err = antreav1alpha2.AddToScheme(scheme)
+	Expect(err).NotTo(HaveOccurred())
+
+	err = runtimev1alpha1.AddToScheme(scheme)
+	Expect(err).NotTo(HaveOccurred())
+
+	// setup kubeCtl.
+	kubeconfig = flag.Lookup("kubeconfig").Value.(flag.Getter).Get().(string)
+	kubeCtl, err = utils.NewKubeCtl(kubeconfig)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(kubeCtl).ToNot(BeNil())
+
+	// set k8sClient.
+	k8sClient, err = utils.NewK8sClient(scheme, "")
+	Expect(err).ToNot(HaveOccurred())
+	Expect(k8sClient).ToNot(BeNil())
+
+	By("Check cert-manager is ready, may wait longer for docker pull")
+	// Increase the timeout for now to get past CI/CD timeout at this point to see what is causing it.
+	err = utils.RestartOrWaitDeployment(k8sClient, "cert-manager", "cert-manager", time.Second*240, false)
+	Expect(err).ToNot(HaveOccurred())
+	err = utils.RestartOrWaitDeployment(k8sClient, "cert-manager-cainjector", "cert-manager", time.Second*120, false)
+	Expect(err).ToNot(HaveOccurred())
+	err = utils.RestartOrWaitDeployment(k8sClient, "cert-manager-webhook", "cert-manager", time.Second*120, false)
+	Expect(err).ToNot(HaveOccurred())
+
+	By("Check antrea controller is ready, may wait longer for docker pull")
+	err = utils.RestartOrWaitDeployment(k8sClient, "antrea-controller", "kube-system", time.Second*120, false)
+	Expect(err).ToNot(HaveOccurred())
+
+	By("Check nephe controller is not running")
+	err = utils.CheckIfConfigured(k8sClient, "nephe-controller", "nephe-system")
+	Expect(err).To(HaveOccurred())
+
+	By("Check VM VPCs are ready")
+	cloudVPC, err = utils.NewCloudVPC(runtimev1alpha1.CloudProvider(cloudProvider))
+	Expect(err).ToNot(HaveOccurred())
+	if !cloudVPC.IsConfigured() {
+		err := cloudVPC.Reapply(time.Second*600, false)
+		Expect(err).ToNot(HaveOccurred())
+		logf.Log.Info("VM VPCs created", "Provider", cloudProvider, "VPCID", cloudVPC.GetVPCID())
+	}
+
+})
+
+var _ = AfterSuite(func() {
+	if preserveSetup == true || kubeCtl == nil {
```

Comment:
What's the difference between `preserveSetup` and `preserveSetupOnFail`?

Reply:
Think of preserveSetupOnFail as config. and preserveSetup as runtime variable. preserveSetup will be set and reset in each test suite. This will indicate to upgrade suite, whether to delete the VM cluster or keep it.